### ISSUE TITLE
Let's add a "first-timers only" label... and use it.

### ISSUE THREAD

**Issue by kytrinyx** [exercism/discussions#8]
Two things recently had me thinking about this. One was a conversation with @petertseng, the other is an idea that is described on the site http://www.firsttimersonly.com

I think that especially with the issues that are about improving cross-track consistency, there are a lot of little issues that we can specifically leave for new contributors. This might mean that it takes a little longer to get them closed, but on the other hand it might actually make it easier for new contributors to jump in and submit changes.

If we have this, we could let the whole CodeNewbie community know, and they've got a bunch of really enthusiastic, eager, and friendly people who are making the shift into development as a profession.

We could also ping the https://twitter.com/yourfirstpr account on twitter.

**Comment by kotp** [comment]
I know I don't have access to add labels in xruby but there are a few that could be useful.

I think the label for firsttimersonly should be added as soon as possible.

**Comment by kytrinyx** [comment]
@kotp I thought if you had commit you had access to labels :/

I think I'll write a script to create the first-timers only label to all the repositories.

**Comment by petertseng** [comment]
I'm of the belief that I can create labels on a reposity that I only have write (and not admin) access to. I believe this because a "New label" button appears on https://github.com/exercism/xgo/labels for me.

**Comment by kytrinyx** [comment]
Yeah, I think I have another script that I wrote the other day that could be repurposed for this.

**Comment by kotp** [comment]
Access to add labels that already exist, but not to create a new label.  That is in another area, settings for the repository itself, and beyond my pay grade.

**Comment by kytrinyx** [comment]
Weird. OK, well I don't mind adding them.

**Comment by kotp** [comment]
I know, right, hard to imagine anything beyond my pay grade. :money_with_wings:

**Comment by kytrinyx** [comment]
@kotp on this page https://github.com/exercism/xruby/labels you don't have a "create" button?

**Comment by kotp** [comment]
I do.  Let me see if I can figure out how you got to that link from the issues screen... found it.... now what color should we use?

**Comment by kytrinyx** [comment]
I'm thinking #159818 which is the green that the default "help wanted" label has when you create a new repository.

**Comment by kytrinyx** [comment]
I've put on my list to script this today, by the way.

**Comment by petertseng** [comment]
Some tl;dr points in bold

At first I was going to say "You'll need to sell me on this - I'd be more convinced of the value of this if we actually had a decent influx of new contributors". But then I realized **maybe the reason we don't get many of them is we haven't been welcoming enough**. And one way to be welcoming is to have suggestions of tasks that they can do and guidance on how to do them. It would be good to write issues for tracks in such a way so as to be understandable to those who aren't necessarily familiar with the tracks for this reason.

For me personally I was a bit reluctant to file easy tasks because part of me thought "So if I can go to the trouble of filing the issue, why don't I just do the task myself?" but **allowing others to take part seems like it has the long-term advantage**. And you can approach it with the mindset of "This allows me to focus on the thornier issues that wouldn't necessarily be as understandable to newcomers"

Another thing I used to think was "It's good to do the easy tasks as fast as I can so that people will think 'Oh, this track stays on top of its issue queue because the number of outstanding issues is small and issues get closed quickly!'". I'm not sure what the right balance to strike there is, but @kytrinyx suggested that as long as we are merging PRs quickly and discussing things there'll be no danger of us not seeming on top of things.

If there are tasks that are super urgent, surely we could do those as soon as we can. But I'm sure there are also a bunch of tasks that are nice-to-have but not absolutely essential. There's a balance to strike here too because I think as a newcomer I would not feel motivated to work on an issue that wouldn't make a difference at all. So it's also good to **lead with an explanation of why the issue is important**.

Basically the theme is letting people know their contributions are welcome, having suggestions for what that contribution could be, and, importantly, letting them know the contribution is valued.

**Comment by kytrinyx** [comment]
> maybe the reason we don't get many of them is we haven't been welcoming enough

I suspect that this is the case.

I would absolutely love to make it so pretty much anyone with just a little bit of programming experience could jump in and successfully contribute. Lots of clear, well-defined, isolated tasks seems like this could make it possible.

If we can show that we consistently do this, then I'm pretty sure I could find a lot of potential folks who have the necessary skills, but no experience with open source. CodeNewbies, the various developer training programs and bootcamps, etc. I think we could get a bunch of people excited about working on this.

As you point out, this would really let us focus in on thornier and harder-to-tackle issues.

I spend a ton of time doing "shallow" work on Exercism (probably 10-15 hours a week on average). Looking over issues, pull requests, minor fixes and updates, triage, support, answering emails. I won't ever have zero of this stuff, but I do think that if you're new to programming, a lot of the stuff that is shallow for me would be great learning experiences for someone else.

> It would be good to write issues for tracks in such a way so as to be understandable to those who aren't necessarily familiar with the tracks for this reason.

Yeah, I was talking to @masters3d about this today. Some of the Objective-C issues are a bit cryptic. It would be hard to jump in as an outsider. And I *know* most of the issues on the main repo are cryptic :/ It's so easy for me to open an issue and say "we should do X", but I forget to say why I think it's important, and where X lives in the codebase, and how we might go about fixing it. I really want to get better at writing these things up in ways that make them easier for other people to tackle.

**Comment by kytrinyx** [comment]
I'm running the script to create the label (both "first-timers only" and "good first patch", which has a slightly different flavor).

I'd love to keep discussing ways in which we could get better at welcoming new contributors, so I'm going to leave this open.

**Comment by kotp** [comment]
@kytrinyx Where should we document the specifics for these two tags?  So that we can understand how to apply these in a meaningful and consistent way?

**Comment by kytrinyx** [comment]
Let's do this in the x-common contributing guide: https://github.com/exercism/x-common/blob/master/CONTRIBUTING.md

**Comment by kytrinyx** [comment]
It's been a few weeks, and I wanted to report back: I've been using the "good first patch" label a lot, but not the "first-timers only" label. The difference is that if we require it to be someone's first ever PR, I worry that we'll have a ton of issues that never get tackled, whereas "good first patch" suggests that it's fairly straight-forward to tackle, but we're not going to get upset if someone deals with it, even if they've made a contribution before.

**Comment by jtigger** [comment]
I see these: https://github.com/search?q=label%3Afirst-timers-only&type=Issues and wonder how many of them just get done in Exercism?

My gut feeling is that those things *are* being tackled by current contributors.  And that it seems "obvious" that it's way more work than it's worth to create an issue for simple things.

I think that actually, the value of bringing on more folks is worth the effort.... but not if we have to craft these issues from scratch each time.  Hoodie seems to have a nice one:

https://github.com/hoodiehq/hoodie-account-server/issues/122

**Comment by kytrinyx** [comment]
> My gut feeling is that those things are being tackled by current contributors.

@jtigger I think that's probably correct.

> I think that actually, the value of bringing on more folks is worth the effort

Yeah, I think that within the exercism.io website repo first timers only issues don't make sense, just because the repo is a disaster. For each of the language tracks, it does make sense, but we would need to know that we have a stream of new contributors. Catch 22, I guess.

**Comment by jtigger** [comment]
On the Java track, I have been specifically trading off the time spent fixing things with creating [issues to fix to fix a small chunk](https://github.com/exercism/xjava/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+patch%22).

My current intent is to groom a perennial backlog of such issues so that we always have between 10-15 of such chunks of work.  If the backlog grows passed 15 such issues, flip and start fixing stuff.  If it drops below 10 such issues, flip and start creating issues.  In this way, work is getting done but there's always surface area for newcomers to jump in.

**Comment by kytrinyx** [comment]
@jtigger I love how that gives a measure of when to do what. Very actionable.

**Comment by kotp** [comment]
Regarding a message from [25 April](https://github.com/exercism/discussions/issues/8#issuecomment-214176249): I am playing with a set of [template files for issues/pull requests](https://github.com/exercism/xruby/issues/497) in the Ruby track, to keep it isolated, and gauge what to put in there and if it is deemed helpful for us, but on a lighter impact, as it affects currently only one track, and it seems like the labels may be mentioned there as a reminder.

**Comment by kytrinyx** [comment]
We've done a great job, I think, with the "good first patch" label.

I'm going to close this issue (thanks everyone!).
